### DISCUSSION TITLE
Pickle-based non-blocking collectives

### DISCUSSION THREAD

**Discussion by Danyc0** [mpi4py/mpi4py#506] [Q&A] [answered]
It looks like there are Python object versions of the non-blocking Isend and Irecv (isend and irecv respectively), and Python object versions of the blocking functions Alltoall, Scatter, Gather and Bcast (Ialltoall, Iscatter, Igather and Ibcast), so is there some fundamental limitation that means there can't be Python object versions of the non-blocking functions (i.e. ialltoall, iscatter, igather, ibcast, etc) ?

**Comment by dalcinl** [comment]
Yes, there are rather basic limitations: 1) There is no "probe" operation for collectives 2) The recv buffer arguments require  knowing in advance the EXACT count of the matching buffer argument in the sending side. These two limitations play together against an (efficient) implementation of non-blocking pickle-based collectives.

A possible workaround users could attempt on their own side is to mix threads + blocking collectives (on a dup communicator for safety) + generalized requests (as the inter-thread signaling mechanism).

Another workaround would be to use the uppercase non-blocking collectives (Ibcast) on a know-size buffer large enough to store the pickled object. Then, at the sending side, you first memcpy the pickled stream in the beginning of buffer, and in the receiving side you unpickle. Obviously, this approach is rather inefficient, as you are communicating unneeded bytes. Nonetheless, it could be good enough if the object to communicate have a small memory footprint, practicality beats purity.

**Comment by dalcinl** [comment, marked as answer]
Update: After merging #507, my previous suggestion related to using generalized requests allows for a more or less straightforward and general implementation. We can use a high-order function to create a non blocking wrapper for any blocking function,with the blocking function blocking on another thread, possibly within a `ThreadPoolExecutor` task. Needless to say, this requires MPI initialized with `MPI_THREAD_MULTIPLE ` support (which is mpi4py's default way to initialize MPI).

With some extra work creating a `Request` subclass adding the lowercase `complete()` method and overriding `wait()`/`test()` the code below may still kind of work. I'll release mpi4py in a couple weeks, so the  effort for backward compatibility is most likely not worth it.

```python
import functools
import threading
import concurrent.futures
from mpi4py import MPI


def nonblocking(function, executor=None):
    """Create non blocking version of blocking function."""
    @functools.wraps(function)
    def wrapper(*args, **kwds):
        request = MPI.Grequest.Start()
        def task():
            try:
                result = function(*args, **kwds)
                request.complete(result)
            except BaseException as exception:
                request.complete(exception)
        if executor is not None:
            executor.submit(task)
        else:
            threading.Thread(target=task).start()
        return request
    return wrapper


if __name__ == "__main__":
    from mpi4py.util.sync import Sequential  # just for ordered print()

    comm = MPI.COMM_WORLD

    ibcast = nonblocking(MPI.Comm.bcast)
    request = ibcast(comm, {comm.rank}, root=comm.size-1)
    result = request.wait()
    with Sequential(comm):
        print(f"[{comm.rank}]:", result, flush=True)

    executor = concurrent.futures.ThreadPoolExecutor()
    iallgather = nonblocking(MPI.Comm.allgather, executor=executor)
    request = iallgather(comm, {comm.rank})
    result = request.wait()
    with Sequential(comm):
        print(f"[{comm.rank}]:", result, flush=True)
```

DISCLAIMER: mpi4py+multithreading has not been stress-tested enough in production, at least from my side. In case of doubt of bad/buggy behavior, try to use the code above with duplicate communicators.

@Danyc0 Please let me know what do you think. This new entry may be the most appropriate answer for your original question.

**Reply by Danyc0** [reply]
I haven't given it a try yet, but looking at your code it looks like a really good work around for what I'm looking for, thank you!
